zirgen/circuit/bigint/bigint2c.cpp

```diff
@@ -0,0 +1,506 @@
+                              APInt::getOneBitSet(256, 4);
+const APInt secp256k1_a(8, 0);
+const APInt secp256k1_b(8, 7);
+/*
+// Base point
+const APInt
+    secp256k1_G_x(256, "79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798", 16);
+const APInt
+    secp256k1_G_y(256, "483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8", 16);
+const APInt
+    secp256k1_order(256, "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141", 16);
+*/
+
+int kArenaConst = 28; // Reg T3 = x28
+uint32_t kArenaTmp = 2;    // Reg SP = x2
+
+struct PolyAtom {
+  uint32_t arena;
+  uint32_t offset;
+  uint32_t size;
+  bool doWrite;
+  PolyAtom() : arena(0), offset(0), size(0) {}
+  PolyAtom(uint32_t arena, uint32_t offset, uint32_t size, bool doWrite)
+      : arena(arena), offset(offset), size(size), doWrite(doWrite) {}
+  bool operator<(const PolyAtom& rhs) const {
+    if (arena != rhs.arena) {
+      return arena < rhs.arena;
+    }
+    if (offset != rhs.offset) {
+      return offset < rhs.offset;
+    }
+    return size < rhs.size;
+  }
+};
+
+struct PolyProd {
+  std::map<PolyAtom, unsigned> atoms;
+  PolyProd() {}
+  PolyProd(const PolyAtom& atom) { atoms[atom] = 1; }
+  bool operator<(const PolyProd& rhs) const { return atoms < rhs.atoms; }
+  size_t degree() const {
+    size_t tot = 0;
+    for (const auto& atom : atoms) {
+      tot += atom.second;
+    }
+    return tot;
+  }
+  PolyProd operator*(const PolyProd& rhs) const {
+    PolyProd out = *this;
+    for (const auto& atom : rhs.atoms) {
+      out.atoms[atom.first] += atom.second;
+    }
+    return out;
+  }
+};
+
+struct Polynomial {
+  std::map<PolyProd, int> terms;
+  Polynomial() {}
+  Polynomial(PolyProd rhs) { terms[rhs] = 1; }
+  size_t degree() const {
+    size_t max = 0;
+    for (const auto& term : terms) {
+      max = std::max(max, term.first.degree());
+    }
+    return max;
+  }
+  void normalize() {
+    for (auto it = terms.begin(); it != terms.end();) {
+      if (it->second == 0) {
+        it = terms.erase(it);
+      } else {
+        ++it;
+      }
+    }
+  }
+  Polynomial operator+(const Polynomial& rhs) const {
+    Polynomial out = *this;
+    for (const auto& term : rhs.terms) {
+      out.terms[term.first] += term.second;
+    }
+    out.normalize();
+    return out;
+  }
+  Polynomial operator-(const Polynomial& rhs) const {
+    Polynomial out = *this;
+    for (const auto& term : rhs.terms) {
+      out.terms[term.first] -= term.second;
+    }
+    out.normalize();
+    return out;
+  }
+  Polynomial operator*(const Polynomial& rhs) const {
+    Polynomial out;
+    for (const auto& lhsTerm : terms) {
+      for (const auto& rhsTerm : rhs.terms) {
+        auto prod = lhsTerm.first * rhsTerm.first;
+        out.terms[prod] += lhsTerm.second * rhsTerm.second;
+      }
+    }
+    out.normalize();
+    return out;
+  }
+};
+
+// TODO: Maybe don't copy;
```

Comment:
I'm a little unclear what you mean with this TODO, maybe clarify comment or fix?

Reply:
Left over from some code I got from jbruestle - doesn't seem to be applicable anymore - I'll remove it.